mediachain/transactor/blockchain_follower.py

```diff
@@ -65,48 +68,82 @@ def cancel(self):
         self._cancelled = True
         self.block_ref_queue.put('__abort__')
         self.incoming_event_queue.put('__abort__')
+        # setting these events will cause the __abort__ messages to be
+        # pulled from the queues.
+        self.catchup_begin.set()
+        self.catchup_complete.set()
+
+    def _clear_queues(self):
+        with self.block_ref_queue.mutex:
+            self.block_ref_queue.queue.clear()
+        with self.block_replay_queue.mutex:
+            self.block_replay_queue.queue.clear()
+        with self.incoming_event_queue.mutex:
+            self.incoming_event_queue.queue.clear()
 
     def _perform_catchup(self):
         if not self.should_catchup:
             return
 
         while True:
+            # block until the event consumer thread tells us to start the
+            # blockchain catchup process
+            self.catchup_begin.wait()
+
             ref = self.block_ref_queue.get()
             if ref == '__abort__':
                 return
 
+            if ref_base58(ref) == self.last_known_block_ref:
+                print('hit last known block: {}'.format(
+                    self.last_known_block_ref
+                ))
+                self.catchup_complete.set()
+                continue
+
             block = self.cache.get(ref)
 
             if block is None:
+                # FIXME: we need to handle this better
```

Comment:
not necessarily; the cache will pull from its "primary" datastore if we don't have it locally, which will be the case when we're doing the catch up for the first time.  Since the primary datastore is the dynamo-backed rpc service, that will already be wrapped in a backoff / retry, and hopefully we'll have added backoff / retry to the IpfsDatastore before we start using it for storing our blocks.

So transient errors should be okay; I'm mostly thinking that if the block just doesn't exist, we should raise an exception or otherwise indicate failure to the consumer.